tests/test_web_poet_rules.py

```diff
@@ -0,0 +1,1173 @@
+        return {"msg": "multiple rule page"}
+
+
+@inlineCallbacks
+def test_multiple_rules_single_page_object() -> None:
+    """A single PO could be used by multiple other rules."""
+    item, deps = yield crawl_item_and_deps(FirstPage)
+    assert item == {"msg": "multiple rule page"}
+    assert_deps(deps, {"page": MultipleRulePage})
+
+    item, deps = yield crawl_item_and_deps(SecondPage)
+    assert item == {"msg": "multiple rule page"}
+    assert_deps(deps, {"page": MultipleRulePage})
+
+    # Calling the replacement should also still work
+    item, deps = yield crawl_item_and_deps(MultipleRulePage)
+    assert item == {"msg": "multiple rule page"}
+    assert_deps(deps, {"page": MultipleRulePage})
+
+
+@attrs.define
+class Product:
+    name: str
+
+
+@handle_urls(URL)
+class ProductPage(ItemPage[Product]):
+    @field
+    def name(self) -> str:
+        return "product name"
+
+
+@inlineCallbacks
+def test_basic_item_return() -> None:
+    """Basic item use case.
+
+    If an item class is asked for and it's available in some rule's ``to_return``
+    parameter, an item class's instance shall be produced by the page object
+    declared inside the rule's ``use`` parameter.
+    """
+    item, deps = yield crawl_item_and_deps(Product)
+    assert item == Product(name="product name")
+    assert_deps(deps, {"item": Product})
+
+    # calling the actual page object should also work
+    item, deps = yield crawl_item_and_deps(ProductPage)
+    assert item == Product(name="product name")
+    assert_deps(deps, {"page": ProductPage})
+
+
+@attrs.define
+class ProductFromParent:
+    name: str
+
+
+@handle_urls(URL)
+class ParentProductPage(ItemPage[ProductFromParent]):
+    @field
+    def name(self) -> str:
+        return "parent product name"
+
+
+@handle_urls(URL)
+class SubclassProductPage(ParentProductPage):
+    @field
+    def name(self) -> str:
+        return "subclass product name"
+
+
+@inlineCallbacks
+def test_item_return_subclass() -> None:
+    """A page object should properly derive the ``Return[ItemType]`` that it
+    inherited from its parent.
+
+    In this test case, there's a clash for the ``url_matcher.Patterns`` since
+    they're exactly the same. This produces a warning message. For conflicts
+    like this, scrapy-poet follows the first ``ApplyRule`` it finds inside the
+    ``SCRAPY_POET_OVERRIDES`` setting.
+
+    To remove this warning, the user should update the priority in
+    ``url_matcher.Patterns`` which is set in ``ApplyRule.for_patterns``.
+    """
+
+    # There should be a warning to the user about clashing rules.
+    rules = [
+        ApplyRule(URL, use=ParentProductPage, to_return=ProductFromParent),
+        ApplyRule(URL, use=SubclassProductPage, to_return=ProductFromParent),
+    ]
+    msg = f"Consider updating the priority of these rules: {rules}"
+
+    with warnings.catch_warnings(record=True) as caught_warnings:
+        item, deps = yield crawl_item_and_deps(ProductFromParent)
```

Comment:
As per https://github.com/scrapinghub/web-poet/pull/112/files#diff-22f478b54a07b372ea47ead887ca01488362ad501692903a9d303511f9d1fe24R157-R163, it would be during import time when the rules are being added to the registry:

```python
from web_poet import consume_modules, default_registry

consume_modules("package_A", "package_B")

# user warning above since:
# 1. @handle_urls are being executed
# 2. which calls add_rule()
# 3. which in turn checks and raises the user warning
```

The tests are set up in a way such that the registry is created each time, alongside the spider via `crawl_item_and_deps()`. This allows us to test each case individually with abstracted setups.